libraries/lib-sentry-reporting/SentryReport.cpp

```diff
@@ -0,0 +1,430 @@
+/*!********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file SentryReport.cpp
+ @brief Define a class to report errors to Sentry.
+
+ Dmitry Vedenko
+ **********************************************************************/
+
+#include "SentryReport.h"
+
+#include <chrono>
+#include <cstring>
+#include <mutex>
+
+#include <algorithm>
+#include <cctype>
+
+#include <rapidjson/document.h>
+#include <rapidjson/writer.h>
+#include <rapidjson/prettywriter.h>
+
+#include <wx/platinfo.h>
+#include <wx/log.h>
+
+#include "CodeConversions.h"
+#include "Uuid.h"
+
+#include "IResponse.h"
+#include "NetworkManager.h"
+
+#include "SentryRequestBuilder.h"
+
+namespace audacity
+{
+namespace sentry
+{
+namespace
+{
+//! This small class is a thread safe store for the information 
+//! we want to add to the exception before the exception occurs.
+//! For example, we may log SQLite3 return codes here, as otherwise 
+//! they wont be available when everything fails
+class ExceptionContext final
+{
+public:
+    //! Adds a new item to the exception context
+   void Add(std::string parameterName, std::string parameterValue)
+   {
+      std::lock_guard<std::mutex> lock(mDataMutex);
+      mData.emplace_back(std::move(parameterName), std::move(parameterValue));
+   }
+
+   //! Return the current context and reset it
+   std::vector<ExceptionData> ConsumeParameters()
+   {
+      std::lock_guard<std::mutex> lock(mDataMutex);
+
+      std::vector<ExceptionData> emptyVector;
+
+      std::swap(mData, emptyVector);
+
+      return emptyVector;
+   }
+
+   //! Get an instance of the ExceptionContext
+   static ExceptionContext& Get()
```

Comment:
Should the constructor be private?

Is it enough to have one global or could there be a need for thread_local ?

Reply:
>Is it enough to have one global or could there be a need for thread_local

It is internationally not thread_local, with the synchronized state.

Reply:
Of course I see the synchronization, so it's not thread_local, but I'm wondering whether it might need to be.

Maybe (very very improbably) there are exceptions on two threads at once and you don't want to combine their data into one report.